File: 90.subsets-ii.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        ans.push_back({});
        for (int i = 0; i < nums.size(); i++) {
            int n = ans.size();
            for (int j = 0; j < n; j++) {
                auto a = ans[j];
                a.push_back(nums[i]);
                ans.push_back(a);
            }
            int k = 1;
            while (i + 1 < nums.size() && nums[i + 1] == nums[i]) {
                for (int j = 0; j < n; j++) {
                    auto a = ans[k * n + j];
                    a.push_back(nums[i]);
                    ans.push_back(a);
                }
                k++;
                i++;
            }
        }
        return ans;
    }
};

File: 90.subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ans = [[]]
        nums.sort()
        l = 0
        for i in range(len(nums)):
            if i == 0 or nums[i] != nums[i - 1]:
                l = len(ans)
            for j in ans[-l:]:
                ans.append(j + [nums[i]])
        return ans